yolov7/detect_ptg.py

```diff
@@ -73,10 +89,111 @@ def read_image(image_fn, imgsz, stride, device, half):
+
+    :return: This will return a tuple of:
+        [0] Torch device instance the model is loaded onto.
+        [1] The model instance.
+        [2] The model's stride.
+        [3] The **adjusted** image input size to be given to the loaded model.
+    """
+    device = select_device(device)
+    model = attempt_load(weights_fp, map_location=device)  # load FP32 model
+    stride = int(model.stride.max())  # model stride
+    imgsz = check_img_size(img_size, s=stride)  # check img_size
+    return device, model, stride, imgsz
+
+
+def predict_image(
+    img0: npt.NDArray,
+    device: torch.device,
+    model: torch.nn.Module,
+    # Preprocessing parameters
+    stride: int,
+    imgsz: Union[int, Tuple[int, int]],
+    half: bool,
+    # Prediction parameters
+    augment: bool,
+    det_conf_threshold: float,
+    iou_threshold: float,
+    filter_classes: Optional[List[int]],
+    class_agnostic_nms: bool,
+):
+    """
+    Perform a full prediction step given a raw BGR input image.
+
+    This function results in a generator that yields individual detection
+    attributes as tuples.
+    Each yielded element will be of the format:
+        * [0] xyxy bounding-box 4-tensor (upper-left, lower-right) in input
+          image coordinate space.
+        * [1] float detection confidence score
+        * [2] integer predicted class index (indexing `model.names`)
+
+    :param img0: Original input image matrix in [H, W, C] format.
+    :param img: Pre-processed image as a torch.Tensor. Assumed to be on the
+        same device as the model.
+    :param imgsz: Integer (square size) or [height, width] tuple specifying the
+        target image size to transform to.
+    :param model: The model to perform inference with. Expecting a module with
+        a `forward` method with the same interface as that in
+        `yolov7.models.yolo.Model`.
+    :param stride: Model stride to take into account.
+    :param device: Device the model is located on, so we can move our tensor
+        there as well.
+    :param half: True if the model is using half width (FP16), false if not
+        (FP32).
+    :param augment: If augmentation should be applied for this inference.
+
+    :return: Generator over individual detections.
+    """
+    height, width = img0.shape[:2]
+
+    # Preprocess image
+    img = preprocess_bgr_img(img0, imgsz, stride, device, half)
+
+    # Predict
+    with torch.no_grad():  # Calculating gradients would cause a GPU memory leak
+        pred = model(img, augment=augment)[0]
+
+    # Apply NMS
+    pred_nms = non_max_suppression(pred, det_conf_threshold, iou_threshold,
+                                   classes=filter_classes,
+                                   agnostic=class_agnostic_nms)
+    assert len(pred_nms) == 1, (
+        f"We are only processing one image, so only a list of 1 element "
+        f"should only every be returned from non_max_suppression. "
```

Comment:
every -> ever

Reply:
I'm actually going to remove this block. I learned that prediction *can* return more than one list element because the `augment` option being true will do that.